Use more granular feature checks in Hosting Configuration page

daledupreez opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

## Proposed Changes

This PR tackles a number of changes stemming from some assumptions that were made when building the Hosting Configuration page in Calypso. The core issue was that all Atomic features were assumed to be controlled via the SFTP feature, but the recent introduction of free eCommerce trials invalidates that assumption, as those sites should have access to Atomic features but not SSH or SFTP access. As such, this PR tackles a few things:
 * We split out hosting features into `basic` and `advanced` buckets, with the following features falling into the basic Atomic bucket behind the `WPCOM_FEATURES_ATOMIC` feature:
    - Transferring the site to Atomic (which has come up in some cases where initial transfers failed)
    - Restoring plugins and themes from the plan
    - Clearing the site cache
      * **Note:** We currently don't include PHP version and web server configuration in this bucket, as it's not clear that free trials should get access to that functionality - it would be better if we have a dedicated feature for this, but we don't currently have that.
 * The logic for showing the activation and upsell banners has been updated:
    - If the site doesn't have the Atomic feature, we show the upsell banner
    - If the site is Atomic, is not transferring, and doesn't have advanced features (as identified by the SFTP feature), we show the upsell banner
       * This is a departure from the old logic, as we still want an upsell for sites that don't have full hosting access (which is currently limited to Woo Express trials)
    - Otherwise we show the activation banner (which may be null/empty)
 * We keep the existing card sequence in the main part of the page and the sidebar, but we re-order them such that any cards for unavailable features are pushed to a `FeatureExample` that appears below the features that are available.
   - This makes it easier for users to access the features they do have access to, and if there are unavailable features, the upsell banner updates in #76783 should make it clearer what those features are.

This PR depends on #76783 to ensure we show consistent upsell banner content for all site types.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Unit tests

Review the test cases covered by the unit tests, and verify that the unit tests succeed:
```
yarn test-client client/my-sites/hosting/test/index.js
```

### Manual testing

* Run this branch locally or via Calypso.live using the link in [this comment](https://github.com/Automattic/wp-calypso/pull/76667#issuecomment-1538213633).
* For each of the test cases below, we want to navigate to _Settings_ -> _Hosting Configuration_ (i.e. `/hosting-config/:siteSlug`, and confirm that the expectations are met.

#### Test cases

  - A site on WPCOM Simple that doesn't have Atomic features, i.e. the site is free, or has a Personal or Premium plan
    * You should see the upsell banner and the _Support_ card, and all other feature cards should be obscured
<img width="1080" alt="Screenshot 2023-05-17 at 11 53 29" src="https://github.com/Automattic/wp-calypso/assets/3376401/6d4b74b3-aae9-4cf5-b9bc-d0683bca3b43">

  - A site on WPCOM Simple that can go Atomic, e.g. a Simple site with the WordPress.com Business plan
    * You should see the activation banner that allows you to enable hosting features, and the _Support_ card should be visible. All other cards should be obscured.
<img width="1080" alt="Screenshot 2023-05-17 at 12 48 07" src="https://github.com/Automattic/wp-calypso/assets/3376401/cc06a2c6-8035-47e5-b455-6f8d8bc36b6b">

  - An Atomic site that already has access to all Atomic features - this includes the Business, eCommerce, and paid Woo Express plans
    * You should see all cards, and no upsell or activation banner
<img width="1080" alt="Screenshot 2023-05-17 at 11 53 12" src="https://github.com/Automattic/wp-calypso/assets/3376401/0ca79515-4a55-4447-8fa8-5359323ab2e7"> 

  - An Atomic site with access to limited features, which specifically requires a site with the Woo Express trial plan added -- you can create such a site by working through the flow from https://woocommerce.com/start.
    * You should see the upsell banner, the _Restore plugins and themes_ card, the _Cache_ card, the _Site backup card, and the _Support_ card. All other cards in the main section should be obscured.
<img width="1080" alt="Screenshot 2023-05-17 at 11 52 56" src="https://github.com/Automattic/wp-calypso/assets/3376401/d7b100d3-65d0-4636-bf8e-244ac64547b5">

  - Use the Atomic revert tool to revert your Woo Express trial site to WordPress.com Simple.
  - Once the revert completes, navigate to _Settings_ -> _Hosting Configuration_
  - Verify that you are shown the activation banner which allows you to request activation of the site's Atomic features
<img width="1080" alt="Screenshot 2023-05-17 at 12 14 38" src="https://github.com/Automattic/wp-calypso/assets/3376401/f77bc446-c021-4ce7-92e7-707e32513b3a">

  - Click on the "Activate" CTA in the banner, and proceed with the transfer
  - Verify that you see the "Please wait" message in the activation banner while the site transfers to Atomic
<img width="1080" alt="Screenshot 2023-05-17 at 12 14 57" src="https://github.com/Automattic/wp-calypso/assets/3376401/e08bd3e9-8867-42e2-b4ab-8dd9deb39135">


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [N/A] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [N/A] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

daledupreez committed:
Use more granular feature checks in Hosting Configuration page

daledupreez committed:
Update banner logic

danielbachhuber reviewed (commented):
> We need to test the following cases:

Can we add some JavaScript tests for these scenarios?

p-jackson commented on client/my-sites/hosting/main.js:
```diff
@@ -143,27 +147,28 @@ class Hosting extends Component {
+					isBasicHostingDisabled &&
+					( englishLocales.includes( locale ) ||
+						i18n.hasTranslation( 'Please activate your hosting access.' ) )
+						? translate( 'Please activate your hosting access.' )
+						: translate( 'Please activate the hosting access to begin using these features.' );
```
IMO the extra case—taking a WooExpress site Atomic—doesn't warrant its own copy. 1) it doesn't seem that likely 2) "these features" doesn't feel like it implies the SFTP to me at least, it's more implies the "advanced" features mentioned above in the sub title.

![woo express that is simple](https://github.com/Automattic/wp-calypso/assets/1500769/23cc8c3e-9011-49d2-a127-e6f77a6d356e)

daledupreez replied:
Agreed.

daledupreez replied:
And the one case where it would occur is for stuck transfers, which we have started tripping over. So this is more likely to be something seen by HEs, but they still need a path for re-triggering a transfer.

p-jackson reviewed (commented):
> Can we add some JavaScript tests for these scenarios?

This'd be great! Might be a heavy lift because it doesn't look like there are any RTL tests for this page yet, so there'll be a bunch of up-front mocking to do. But there are so many permutations of this page, it'd be great to get coverage of them without having to manually set up all the situations.

> We need to test the following cases

I've tested these cases and things are working as expected.

This case where the WooExpress trial site is on Atomic doesn't look great IMO. The main issue for me is the gradient. Would it look better to just have a solid greyed out look rather than something that fades to transparent?

![CleanShot 2023-05-10 at 11 40 11@2x](https://github.com/Automattic/wp-calypso/assets/1500769/aba85332-0512-436e-bc7a-3237ea6f2594)

daledupreez committed:
Refactor code to show available feature cards first

daledupreez commented:
Thanks for the feedback, @danielbachhuber and @p-jackson!

I realised that a big part of my broader confusion was due to us using the older upsell content/style when the site has an eCommerce trial plan. When we implemented an alternate upsell path for those plans, we used the older nudge, as you were still experimenting with the now-default upsell. I've split out #76783 to use the new style, and will keep iterating on this PR so we do the following:
* Show the activation banner to transfer to Atomic when the site has the base Atomic feature, but is not Atomic, or is transferring to Atomic
* Refactor the code so we show all available cards at the top of the screen, and keep all unavailable feature cards behind the `FeatureExample` block
  - The core thinking here is that if the user has basic Atomic features, the only things they can access are the plugin restore and clear cache features
  - While the web server settings are technically a general platform feature, I don't know that we want folks with plain Atomic features to manipulate their PHP version and other web server settings

daledupreez commented:
> > We need to test the following cases:
>
> Can we add some JavaScript tests for these scenarios?

Absolutely! I just wanted to explore the approach and get some feedback before writing tests.

danielbachhuber commented:
> I've split out https://github.com/Automattic/wp-calypso/pull/76783 to use the new style, and will keep iterating on this PR so we do the following:

Cool, sounds good.

> While the web server settings are technically a general platform feature, I don't know that we want folks with plain Atomic features to manipulate their PHP version and other web server settings

We may need to keep the PHP version switcher available, so they can easily roll back when a new PHP version is available, etc.

daledupreez commented on client/my-sites/hosting/main.js:
```diff
@@ -53,6 +54,125 @@ import WebServerSettingsCard from './web-server-settings-card';
+	isWpcomStagingSite,
+	siteId,
+} ) => {
+	const mainCards = useMemo( () => {
```
Is this unnecessary? I am thinking that I'm mostly depending on the incoming props, which means React should (hopefully) identify that the component's props are unchanged.

p-jackson replied:
I'm pretty sure React re-renders the component whenever the parent re-renders; regardless of whether the props have changed (unless you use `React.memo`). But the `useMemo` still isn't necessary IMO. The filtering logic doesn't look expensive, so unless there's a measurable performance impact I think the logic is clearer without pre-emptively memoising things.

daledupreez committed:
Use original wording for activation message

daledupreez committed:
Rename CardsByEnabledFeatures to ShowEnabledFeatureCards

daledupreez committed:
Only add a FeatureExample wrapper when needed

daledupreez committed:
Improve logic for activation notice

daledupreez committed:
Add cases where we don't show the upgrade banner

daledupreez committed:
Add unit/integration tests

daledupreez committed:
Remove useMemo() wrappers around cards

daledupreez committed:
Stub out staging site cards

sixhours commented:
I went through the testing instructions and the upsell card does *not* show on a fresh Woo Express Trial site until after I've reverted it and restored it to Atomic.

Upon purchasing the Trial and navigating to `/hosting-config/siteSlug`:

<img width="1664" alt="Screen Shot 2023-06-01 at 9 56 44 AM" src="https://github.com/Automattic/wp-calypso/assets/2124984/7c1c953a-d284-4ea8-a1be-ec0e556980f0">

After reverting to Simple, then activating hosting features to transfer the site back to Atomic:

<img width="1669" alt="Screen Shot 2023-06-01 at 10 04 24 AM" src="https://github.com/Automattic/wp-calypso/assets/2124984/02db29b4-8097-4916-956f-2040debbdc8a">

daledupreez commented:
Thanks for the review, @sixhours!

I dug into that specific behaviour for the newly-created Woo Express trial site, and I found that this is a situation that is almost certainly transient, and is linked to the site not being fully synched yet.

I believe the relevant code is here:
https://github.com/Automattic/wp-calypso/blob/b60e83c34d848e3b4cbc977cc062c0e4069216fc/client/blocks/upsell-nudge/index.jsx#L79-L93

This is the component that we use for the upsell nudge, albeit via the `HostingUpsellNudge` wrapper component.

I _think_ we can move forward with the patch, as that behavour seems to be a timing issue that will resolve after a few minutes.

sixhours reviewed (approved):
In that case, 🚢 it! :D
